fainder/execution/parallel_processing.py

```diff
@@ -0,0 +1,247 @@
+
+    # Filter histograms by ID
+    filtered_hists: list[tuple[int | np.integer[Any], Histogram]] = [
+        (id_f, _worker_state.hists[id_f]) for id_f in id_filter if id_f in _worker_state.hists
+    ]
+
+    # Process the histograms
+    return np.fromiter(
+        (
+            np.uint32(id_)
+            for id_, hist in filtered_hists
+            if query_histogram(hist, estimation_mode="over", query=query, density=True)
+        ),
+        dtype=np.uint32,
+    )
+
+
+def partition_histogram_ids(
+    hist_ids: list[int],
+    num_partitions: int,
+    chunk_layout: FainderChunkLayout = FainderChunkLayout.CONTIGUOUS,
+) -> dict[int, set[int]]:
+    """Partition histogram IDs into roughly equal chunks.
+
+    Args:
+        hist_ids: List of histogram IDs to partition
+        num_partitions: Number of partitions to create
+        contiguous: If True, use contiguous chunks; if False, distribute in round-robin fashion
+
+    Returns:
+        Dictionary mapping partition ID to list of histogram IDs
+    """
+    chunks: dict[int, set[int]] = {i: set() for i in range(num_partitions)}
+
+    if chunk_layout == FainderChunkLayout.CONTIGUOUS:
+        # Original contiguous chunking strategy
+        chunk_size = len(hist_ids) // num_partitions
+        remainder = len(hist_ids) % num_partitions
+
+        start_idx = 0
+        for i in range(num_partitions):
+            end_idx = start_idx + chunk_size + (1 if i < remainder else 0)
+            chunks[i] = set(hist_ids[start_idx:end_idx])
+            start_idx = end_idx
+    elif chunk_layout == FainderChunkLayout.ROUND_ROBIN:
+        # Round-robin distribution for more balanced workload
+        current_partition = 0
+        for hist_id in hist_ids:
+            chunks[current_partition].add(hist_id)
+            if current_partition == num_partitions - 1:
+                current_partition = 0
+            else:
+                current_partition += 1
+    else:
+        raise ValueError(f"Unsupported chunk layout: {chunk_layout}")
+
+    return chunks
+
+
+class ParallelHistogramProcessor:
+    """Class for parallel processing of histogram queries."""
+
+    def __init__(
+        self,
+        histogram_path: str | Path,
+        num_workers: int | None = None,
+        num_chunks: int | None = None,
+        chunk_layout: FainderChunkLayout = FainderChunkLayout.CONTIGUOUS,
+    ) -> None:
+        """Initialize the parallel processor with histograms.
+
+        Args:
+            histogram_path: Path to the histogram file or base file path for split files
+            num_workers: Number of worker processes to use. If None, uses CPU count - 1.
+            num_chunks: Number of chunks to split the histograms into. If None, uses num_workers.
+            contiguous: If True, use contiguous chunks of histograms;
+                        if False, distribute in round-robin fashion
+        """
+        self.num_workers = (num_workers or os.cpu_count() or 2) - 1
```

Comment:
Why do you subtract 1 from num_workers? Isn't it confusing if the user requests 5 workers and only gets 4?